state-chain/chains/src/refund_parameters.rs

```diff
@@ -0,0 +1,255 @@
+	address::{AddressConverter, EncodedAddress, ForeignChainAddress, IntoForeignChainAddress},
+	ccm_checker::CcmValidityChecker,
+	CcmChannelMetadataUnchecked, CcmDepositMetadataChecked, CcmDepositMetadataUnchecked, Chain,
+};
+use cf_amm_math::Price;
+use cf_primitives::{Asset, AssetAmount};
+use codec::{Decode, Encode, MaxEncodedLen};
+use frame_support::sp_runtime::DispatchError;
+use scale_info::TypeInfo;
+use serde::{Deserialize, Serialize};
+use sp_std::{cmp::Ord, convert::Into, fmt::Debug, prelude::*};
+
+/// AccountOrAddress is a enum that can represent an internal account or an external address.
+/// This is used to represent the destination address for an egress or an internal account
+/// to move funds internally.
+#[derive(Clone, Debug, PartialEq, Eq, Encode, Decode, TypeInfo, PartialOrd, Ord)]
+pub enum AccountOrAddress<AccountId, Address> {
+	InternalAccount(AccountId),
+	ExternalAddress(Address),
+}
+
+#[derive(
+	Clone,
+	Debug,
+	PartialEq,
+	Eq,
+	Encode,
+	Decode,
+	TypeInfo,
+	MaxEncodedLen,
+	Serialize,
+	Deserialize,
+	PartialOrd,
+	Ord,
+)]
+/// Generic type for Refund Parameters.
+///
+/// The abstract `RefundDetails` represents additional metadata that may be required for refunding
+/// via CCM. Before verification this is an unchecked byte payload.
+///
+/// Avoid constructing this type directly: prefer to use one of the aliases. Usually you start with
+/// `ChannelRefundParametersUnchecked` and then convert it via
+/// `ChannelRefundParametersUncheckedInternalAddress` into `ChannelRefundParametersCheckedInternal`.
+///
+/// Example:
+/// ```ignore
+/// let checked_params = unchecked_refund_params
+///     .with_foreign_chain_refund_address::<Solana>()
+///     .with_checked_metadata(
+///         source_address,
+///         refund_asset
+///     )?;
+/// ```
+pub struct ChannelRefundParameters<A, CcmRefundDetails> {
+	pub retry_duration: cf_primitives::BlockNumber,
+	pub refund_address: A,
+	pub min_price: Price,
+	pub refund_ccm_metadata: CcmRefundDetails,
+}
+
+/// Refund parameters with CCM metadata that has not yet been checked for validity.
+///
+/// Most incoming refund parameters will be of this type.
+pub type ChannelRefundParametersUnchecked<A> =
+	ChannelRefundParameters<A, Option<CcmChannelMetadataUnchecked>>;
+
+/// Refund parameters with CCM metadata that *has* been checked for validity.
+pub type ChannelRefundParametersChecked<A> =
+	ChannelRefundParameters<A, Option<CcmDepositMetadataChecked<ForeignChainAddress>>>;
+
+/// Refund parameters with CCM metadata that has not yet been checked for validity, but where the
+/// refund address has been parsed into the internal `ForeignChainAddress` type.
+pub type ChannelRefundParametersUncheckedInternalAddress =
+	ChannelRefundParametersUnchecked<ForeignChainAddress>;
+
+/// The legacy refund parameters type that does not include any ccm refund metadata.
+pub type ChannelRefundParametersLegacy<RefundAddress> = ChannelRefundParameters<RefundAddress, ()>;
```

Comment:
This legacy type is not actually used anywhere (except for tests). This WAS originally used only in migrations, but not anymore. 
If we want, we do not need to create this type alias

Reply:
We need to keep legacy types of API compatibility, no? I'm pretty sure I'm using this in my PR. Not a problem for me to create it again, but I wouldn't worry about cleaning this PR up tbh, I can do it in my follow-up PR.